### PR TITLE
253 gaps in exports

### LINKED ISSUE #253: Gaps between paragraphs - remove from exports

### PR DESCRIPTION
**Pull request by aniablaziak**
**Is your Pull Request request related to another [issue](https://github.com/bbc/react-transcript-editor/issues) in this repository ?**      

Resolves #253 

**Describe what the PR does**    

Removes gaps from exported files for consecutive transcripts elements. 

**State whether the PR is ready for review or whether it needs extra work**    

I tested exports to JSON, Word, txt files, but need to check in with someone who has the software to verify that other exports work 🙏 

**Additional context**    

To test try adding a few consecutive elements selected together, as well adding consecutive elements one by one. Both should work the same and there should be no gaps between consecutive transcripts.

### CODE REVIEW

**Comment by aniablaziak** [comment]
thanks both! 🙏 just pushed some fixes for that, the exported file _looks_ more like it now 🙂

**Comment by allishultes** [comment]
I can't really test the imports, but looking at the export files they all look good! Is the ticket meant to address ADL as well? This was the only one I wasn't sure adjusted - might need changes to the `formatToADLEvent` function? 

My result for five consecutive clips looked like:

```
<EVENT_LIST>
	(Entry)	1	(Cut)	I	1	1~2	1~2	00:00:16:08/0000	00:00:00:00/0000	00:00:22:09/0000	_	(Rem) NAME ""
	(Entry)	2	(Cut)	I	1	1~2	1~2	00:00:24:08/0000	00:00:22:09/0000	00:00:23:19/0000	_	(Rem) NAME ""
	(Entry)	3	(Cut)	I	1	1~2	1~2	00:00:25:18/0000	00:00:23:19/0000	00:00:23:23/0000	_	(Rem) NAME ""
	(Entry)	4	(Cut)	I	1	1~2	1~2	00:00:25:22/0000	00:00:23:23/0000	00:00:30:10/0000	_	(Rem) NAME ""
	(Entry)	5	(Cut)	I	1	1~2	1~2	00:00:32:09/0000	00:00:30:10/0000	00:00:36:18/0000	_	(Rem) NAME ""
</EVENT_LIST>
```

I'd imagine that'd want to look similarly to the EDL / FCPX, which looks great! But I can't test anything so maybe Sarah or Sevi can try this out and confirm.

**Comment by sarahrainbow** [comment]
So I just exported to FCPX and a few strange things happening…

Initially I noticed that the export wasn't quite long enough (just cut off the last word of my edit) 
I tried re-exporting and noticed the edit had been duplicated. Each time I re-export, it adds another edit onto the end of my edit… if you catch my drift? Is there some kind of += saving into state going on?

File: src/Components/PaperEditor/ProgrammeScriptContainer/ExportDropdown/edl.js [hunk 1]
```diff
@@ -27,17 +27,34 @@ const formatToEDLEvent = (transcript, element) => {
...
-      const edlEvent = formatToEDLEvent(transcript, element);
-      res.index += 1;
-      edlEvent.id = res.index;
-      res.events.push(edlEvent);
-      console.log('res', element, transcript);
+const mergeConsecutiveElements = (elements) => {
+  const paperCuts = elements.filter(element => element.type === 'paper-cut');
+
+  return paperCuts.reduce((accumulator, currentElement) => {
+    const prevElement = accumulator[accumulator.length - 1];
```
Comment (tamsingreen) [inline src/Components/PaperEditor/ProgrammeScriptContainer/ExportDropdown/edl.js:34]: I think the problem is here - if the accumulator is empty then prevElement is ???
I'd change this reduce function body to something like: 
```
    if (accumulator.length > 0) {
      const prevElement = accumulator[accumulator.length - 1];
      const areElementsConsecutive = currentElement.transcriptId === prevElement.transcriptId && prevElement.end === currentElement.start;
      if (areElementsConsecutive) {
        prevElement.end = currentElement.end;
      }
    }
    else {
      accumulator.push(currentElement);
    }

    return accumulator;
```
However!
This code is mutating objects within arrays which are being passed all around this module - lots of potential for things to get messed up, so better to write it immutably, such as:
```
    if (accumulator.length > 0) {
      const prevElement = accumulator[accumulator.length - 1];
      const areElementsConsecutive = currentElement.transcriptId === prevElement.transcriptId && prevElement.end === currentElement.start;
      if (areElementsConsecutive) {
        const mergedElement = { ...prevElement, end: currentElement.end };

        return [ ...accumulator.slice(0, accumulator.length - 1), mergedElement ];
      }
    } else {
      return [ ...accumulator, currentElement ];
    }

Comment (tamsingreen) [reply src/Components/PaperEditor/ProgrammeScriptContainer/ExportDropdown/edl.js:34]: If I add three paragraphs individually or together to a script, the EDL lists them as one clip (I hope that was the right test?)

**Review by sarahrainbow — APPROVED** [review]
all good with me now!

File: src/Components/PaperEditor/ProgrammeScriptContainer/index.js [hunk 2]
```diff
@@ -292,7 +289,8 @@ const ProgrammeScriptContainer = (props) => {
     return totalDuration;
   };
 
-  const formatSingleParagaph = ({ start, end, speaker, transcriptId, words }) => {
+  const formatSingleParagaph = (selection) => {
```
Comment (tamsingreen) [inline src/Components/PaperEditor/ProgrammeScriptContainer/index.js:292]: typo

**Review by tamsingreen — APPROVED** [review]
Code looks good - have suggested a couple of minor tweaks.
I've inspected the exports (EDL, ADL & FCP) and a sequence of clips is treated as one clip. Let's test the Sadie import when it's on Test and Sevi has time to check it.

File: src/Components/PaperEditor/ProgrammeScriptContainer/ExportDropdown/edl.js [hunk 3]
```diff
@@ -27,17 +27,39 @@ const formatToEDLEvent = (transcript, element) => {
...
+    return [ ...accumulator, currentElement ];
+  }, []);
+};
+
+const getEDLSq = (title, elements, transcripts) => {
+  const mergedElements = mergeConsecutiveElements(elements);
+
+  return mergedElements.reduce((res, element) => {
+    const transcript = getCurrentTranscript(element, transcripts);
+    const edlEvent = formatToEDLEvent(transcript, element);
```
[the remark this answers is not in the document]
Comment (sarahrainbow) [reply src/Components/PaperEditor/ProgrammeScriptContainer/ExportDropdown/edl.js:58]: very nice! Have done 👍

**Pull request merged**